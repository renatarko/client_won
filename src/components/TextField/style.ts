import styled, { DefaultTheme, css } from "styled-components";
import { TextFieldProps } from ".";

type IconPositionProps = Pick<TextFieldProps, "iconPosition">;

type WrapperProps = Pick<TextFieldProps, "disabled"> & {
  errorMessage?: boolean;
};

const wrapperModifiers = {
  disabled: (theme: DefaultTheme) => css`
    ${Label}, ${Input}, ${Icon} {
      cursor: not-allowed;
      color: ${theme.colors.gray};

      &::placeholder {
        color: currentColor;
      }
    }
  `,

  errorMessage: (theme: DefaultTheme) => css`
    ${InputWrapper} {
      border-color: ${theme.colors.error};
    }

    ${Icon}, ${Label} {
      color: ${theme.colors.error};
    }
  `,
};

export const Wrapper = styled.div<WrapperProps>`
  ${({ theme, disabled, errorMessage }) => css`
    ${disabled && wrapperModifiers.disabled(theme)}
    ${errorMessage && wrapperModifiers.errorMessage(theme)}
  `}
`;

export const InputWrapper = styled.div`
  ${({ theme }) => css`
    display: flex;
    background: ${theme.colors.lightGray};
    border-radius: 0.2rem;
    padding: 0 ${theme.spacings.xsmall};
    border: 0.2rem solid;
    border-color: ${theme.colors.lightGray};
    &:focus-within {
      box-shadow: 0 0 0.5rem ${theme.colors.primary};
    }
  `}
`;

export const Input = styled.input<IconPositionProps>`
  ${({ theme, iconPosition }) => css`
    color: ${theme.colors.black};
    font-family: ${theme.font.family};
    font-size: ${theme.font.sizes.medium};
    padding: ${theme.spacings.xxsmall} 0;
    padding-${iconPosition}: ${theme.spacings.xsmall};
    background: transparent;
    border: 0;
    outline: none;
    width: 100%;
  `}
`;

export const Label = styled.label`
  ${({ theme }) => css`
    font-size: ${theme.font.sizes.small};
    cursor: pointer;
  `}
`;

export const Icon = styled.div<IconPositionProps>`
  ${({ theme, iconPosition }) => css`
    display: flex;
    width: 2.1rem;
    color: ${theme.colors.gray};
    justify-content: flex-end;
    order: ${iconPosition === "right" ? 1 : 0};

    & > svg {
      width: 100%;
    }
  `}
`;

export const ErrorMessage = styled.p`
  ${({ theme }) => css`
    color: ${theme.colors.error};
    font-size: ${theme.font.sizes.xsmall};
    margin-left: 0.5rem;
  `}
`;
