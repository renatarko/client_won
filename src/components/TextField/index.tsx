import React, { InputHTMLAttributes, useState } from "react";
import * as S from "./style";

export type TextFieldProps = {
  label?: string;
  labelFor?: string;
  valueInicial?: string;
  onInput?: (value: string) => void;
  icon?: React.ReactNode;
  iconPosition?: "right" | "left";
  disabled?: boolean;
  errorMessage?: string;
} & InputHTMLAttributes<HTMLInputElement>;

const TextField = ({
  label,
  labelFor,
  valueInicial = "",
  onInput,
  icon,
  iconPosition = "left",
  disabled = false,
  errorMessage,
  ...props
}: TextFieldProps) => {
  const [value, setValue] = useState(valueInicial);

  const onChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    const newValue = e.currentTarget.value;

    setValue(newValue);

    !!onInput && onInput(newValue);
  };
  return (
    <S.Wrapper disabled={disabled} errorMessage={!!errorMessage}>
      {!!label && <S.Label htmlFor={labelFor}>{label}</S.Label>}
      <S.InputWrapper>
        {!!icon && <S.Icon iconPosition={iconPosition}>{icon}</S.Icon>}
        <S.Input
          type="text"
          onChange={onChange}
          value={value}
          iconPosition={iconPosition}
          disabled={disabled}
          {...props}
        />
      </S.InputWrapper>
      {!!errorMessage && <S.ErrorMessage>{errorMessage}</S.ErrorMessage>}
    </S.Wrapper>
  );
};

export default TextField;
