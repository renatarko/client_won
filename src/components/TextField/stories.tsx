import { Meta, Story } from "@storybook/react/types-6-0";
import { EnvelopeFill } from "styled-icons/bootstrap";
import TextField, { TextFieldProps } from ".";

export default {
  title: "Form/TextField",
  component: TextField,
  args: {
    id: "email",
    label: "E-mail",
    labelFor: "email",
    initialValue: "",
    placeholder: "Digite seu e-mail",
    icon: <EnvelopeFill />,
  },
  argTypes: {
    onInput: { action: "changed" },
    icon: { type: "boolean" },
  },
} as Meta;

export const Default: Story<TextFieldProps> = (args) => (
  <div style={{ width: "50vw" }}>
    <TextField {...args} />
  </div>
);

export const WithError: Story<TextFieldProps> = (args) => (
  <div style={{ width: "50vw" }}>
    <TextField {...args} />
  </div>
);

WithError.args = {
  errorMessage: "Ops... something is wrong",
};
