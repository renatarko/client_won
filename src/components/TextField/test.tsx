import { screen, waitFor } from "@testing-library/react";

import userEvent from "@testing-library/user-event";
import { EnvelopeFill } from "styled-icons/bootstrap";
import { renderWithTheme } from "utils/tests/helpers";
import TextField from ".";

describe("<TextField/>", () => {
  it("Renders with Label", () => {
    renderWithTheme(<TextField label="Field" name="input" />);

    expect(screen.getByLabelText("Field")).toBeInTheDocument();
  });

  it("Renders withou Label", () => {
    renderWithTheme(<TextField />);

    expect(screen.queryByLabelText("Field")).not.toBeInTheDocument();
  });

  it("Renders with Placeholder", () => {
    renderWithTheme(<TextField placeholder="your name" />);

    expect(screen.getByPlaceholderText("your name")).toBeInTheDocument();
  });

  it("should render with icon", () => {
    renderWithTheme(<TextField icon={<EnvelopeFill data-testid="icon" />} />);

    expect(screen.getByTestId("icon")).toBeInTheDocument();
  });

  it("should render icon on the right side", () => {
    renderWithTheme(
      <TextField
        icon={<EnvelopeFill data-testid="icon" />}
        iconPosition="right"
      />
    );

    expect(screen.getByTestId("icon").parentElement).toHaveStyle({ order: 1 });
  });

  it("should render error message", () => {
    const { container } = renderWithTheme(
      <TextField
        icon={<EnvelopeFill data-testid="icon" />}
        iconPosition="right"
        label="TextField"
        name="TextField"
        errorMessage="error message"
      />
    );
    expect(screen.getByText("error message")).toBeInTheDocument();
    expect(container.firstChild).toMatchSnapshot();
  });

  it("Does not changes its value when disabled", async () => {
    const onInput = jest.fn();
    renderWithTheme(
      <TextField
        icon={<EnvelopeFill data-testid="icon" />}
        iconPosition="right"
        disabled
      />
    );
    const input = screen.getByRole("textbox");
    expect(input).toBeDisabled();

    const text = "This is my text";
    userEvent.type(input, text);

    await waitFor(() => {
      expect(input).not.toHaveValue(text);
    });
    expect(onInput).not.toHaveBeenCalled();
  });

  it("Changes its value when typing", async () => {
    const onInput = jest.fn();
    renderWithTheme(
      <TextField label="textField" name="textField" onInput={onInput} />
    );

    const input = screen.getByRole("textbox");
    const text = "This is my text";
    userEvent.type(input, text);

    await waitFor(() => {
      expect(input).toHaveValue(text);
    });
    expect(onInput).toHaveBeenCalledWith(text);
  });
});
