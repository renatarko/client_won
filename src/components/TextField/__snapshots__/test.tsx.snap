// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TextField/> should render error message 1`] = `
.c7 {
  display: inline-block;
  vertical-align: middle;
  overflow: hidden;
}

.c0 .c3 {
  border-color: #FD3500;
}

.c0 .c5,
.c0 .c1 {
  color: #FD3500;
}

.c4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  background: #EAEAEA;
  border-radius: 0.2rem;
  padding: 0 1.6rem;
  border: 0.2rem solid;
  border-color: #EAEAEA;
}

.c4:focus-within {
  box-shadow: 0 0 0.5rem #F231A5;
}

.c8 {
  color: #030517;
  font-family: Poppins,-apple-system,BlinkMacSystemFont,'Segoe UI',Roboto,Oxygen,Ubuntu,Cantarell,'Open Sans','Helvetica Neue',sans-serif;
  font-size: 1.6rem;
  padding: 0.8rem 0;
  padding-right: 1.6rem;
  background: transparent;
  border: 0;
  outline: none;
  width: 100%;
}

.c2 {
  font-size: 1.4rem;
  cursor: pointer;
}

.c6 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  width: 2.1rem;
  color: #8F8F8F;
  -webkit-box-pack: end;
  -webkit-justify-content: flex-end;
  -ms-flex-pack: end;
  justify-content: flex-end;
  -webkit-order: 1;
  -ms-flex-order: 1;
  order: 1;
}

.c6 > svg {
  width: 100%;
}

.c9 {
  color: #FD3500;
  font-size: 1.2rem;
  margin-left: 0.5rem;
}

<div
  class="c0"
>
  <label
    class="c1 c2"
    for="TextField"
  >
    TextField
  </label>
  <div
    class="c3 c4"
  >
    <div
      class="c5 c6"
    >
      <svg
        aria-hidden="true"
        class="c7"
        data-testid="icon"
        fill="currentColor"
        focusable="false"
        viewBox="0 0 16 16"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M.05 3.555A2 2 0 0 1 2 2h12a2 2 0 0 1 1.95 1.555L8 8.414.05 3.555ZM0 4.697v7.104l5.803-3.558L0 4.697ZM6.761 8.83l-6.57 4.027A2 2 0 0 0 2 14h12a2 2 0 0 0 1.808-1.144l-6.57-4.027L8 9.586l-1.239-.757Zm3.436-.586L16 11.801V4.697l-5.803 3.546Z"
        />
      </svg>
    </div>
    <input
      class="c8"
      type="text"
      value=""
    />
  </div>
  <p
    class="c9"
  >
    error message
  </p>
</div>
`;
